src/ol/layer/STAC.js

```diff
@@ -0,0 +1,625 @@
+    }
+
+    if (this.getGeoTIFFSourceOptions_) {
+      options = await this.getGeoTIFFSourceOptions_(options, asset);
+    }
+
+    const tileserverFallback = async (asset, layer) => {
+      if (layer) {
+        this.getLayers().remove(layer);
+      }
+      return await this.addTileLayer_(asset);
+    };
+    try {
+      const source = new GeoTIFF(options);
+      const layer = new WebGLTileLayer({source});
+      if (this.useTileLayerAsFallback_) {
+        const errorFn = () => tileserverFallback(asset, layer);
+        source.on('error', errorFn);
+        source.on('tileloaderror', errorFn);
+        // see https://github.com/openlayers/openlayers/issues/14926
+        source.on('change', () => {
+          if (source.getState() === 'error') {
+            tileserverFallback(asset, layer);
+          }
+        });
+        layer.on('error', errorFn);
+        // Call this to ensure we can load the GeoTIFF, otherwise try fallback
+        await source.getView();
+      }
+      this.addLayer_(layer, asset);
+      return layer;
+    } catch (error) {
+      if (this.useTileLayerAsFallback_) {
+        return await tileserverFallback(asset, null);
+      }
+      this.handleError_(error);
+    }
+  }
+
+  /**
+   * @private
+   * @param {Asset|Link} [data] A STAC Asset or Link
+   * @return {Promise<TileLayer>} Resolves with a TileLayer when complete.
+   */
+  async addTileLayer_(data) {
+    /**
+     * @type {XYZSourceOptions}
+     */
+    let options = {
+      crossOrigin: this.crossOrigin_,
+      url: this.buildTileUrlTemplate_(data),
+    };
+    if (this.getXYZSourceOptions_) {
+      options = await this.getXYZSourceOptions_(options, data);
+    }
+    const layer = new TileLayer({
+      source: new XYZ(options),
+    });
+    this.addLayer_(layer, data);
+    return layer;
+  }
+
+  /**
+   * @param {Layer|LayerGroup} [layer] A Layer to add to the LayerGroup
+   * @param {STAC|Asset} [data] The STAC data source
+   * @param {number} [zIndex=0] The z-index for the layer
+   * @private
+   */
+  addLayer_(layer, data, zIndex = 0) {
+    layer.set('stac', data);
+    layer.setZIndex(zIndex);
+    this.getLayers().push(layer);
+  }
+
+  /**
+   * @private
+   * @return {VectorLayer|null} The vector layer showing the geometry/bbox.
+   */
+  addFootprint_() {
+    let geojson = null;
+    const data = this.getData();
+    if (data.isItemCollection() || data.isCollectionCollection()) {
+      geojson = toGeoJSON(data.getBoundingBox());
+    } else {
+      geojson = data.toGeoJSON();
+    }
+
+    if (geojson) {
+      const features = new GeoJSON().readFeatures(geojson, {
+        featureProjection: 'EPSG:3857', // TODO: revisit this
+      });
+      const source = new VectorSource({features});
+      const vectorLayer = new VectorLayer({
+        source,
+        style: getBoundsStyle(this.boundsStyle_, this),
+      });
+      vectorLayer.set('bounds', true);
+      features.forEach((feature) => feature.set('stac', data));
```

Comment:
A few thoughts here:
* If you configure the source with a `loader` instead of `features`, you can use the view projection, which is passed to the loader as 3rd argument. Then you do not need to hard-code the projection.
* Instead of spreading metadata across assets that the STAC layer group creates, I'd suggest making them available through the STAC instance (e.g. `getBoundsLayer()` instead of setting a `bounds` boolean on the layer. Same for the `stac` data, which should not be set on each feature.

Reply:
Thanks, good pointers!

1. Switched to use a loader, which seems to work perfectly.
2. I removed the stac metadata for the features, this is not needed anymore (due to the change for https://github.com/openlayers/openlayers/pull/14927#discussion_r1282171332). I added a `getBoundsLayer()` method, but the metadata on the assets is still valuable due to the nesting of the LayerGroups, I think. In case you go through all layers, it is helpful to be able to retrieve the relevant STAC entities. Using that for example for customite a LayerSwitcher or for the "click detection" in stacUtils / the STAC Item Collection example.